### ISSUE TITLE
KLqp with Multinomial class return nan value

### ISSUE THREAD

**Issue by YiMX** [blei-lab/edward#724]
I tried using Edward to train a neural classifier with KLqp and I used one-hot labels. The model I defined the output is Multinomial with total_count=1.0 which should be mathematically the same as Categorical. But the  Categorical works fine while the Multinomial does not.

**Comment by dustinvtran** [comment]
Thanks for noting this. Could you give a minimal example reproducing this error?

**Comment by YiMX** [comment]
Please see https://github.com/YiMX/Adversarial-Training/blob/master/Bayesian_MNIST.ipynb 
I include Categorical with comment symbol '#'

**Comment by dustinvtran** [comment]
After some investigating, the issue comes down to instability of Categorical's [`_log_prob`](https://github.com/tensorflow/tensorflow/blob/r1.3/tensorflow/python/ops/distributions/multinomial.py#L258) implementation. Namely, the log-density requires operations on Categorical's probability vector and not the logit parameterization. I'm not sure if the log-density can be made more stable. In any case, contributions to Tensorflow are welcome.

Minimal example built from yours:
```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import edward as ed
import numpy as np
import tensorflow as tf

from edward.models import Multinomial, Normal, Categorical
from tensorflow.examples.tutorials.mnist import input_data

mnist = input_data.read_data_sets('/Users/dvt/data/mnist', one_hot=True)

def forward(X, W_0, b_0, W_1, b_1):
  hidden_layer = tf.nn.relu(tf.matmul(X, W_0) + b_0)
  output = tf.matmul(hidden_layer, W_1) + b_1
  return output

num_features = 784
num_classes = 10
hidden_units = 500

W_0 = Normal(loc=tf.zeros([num_features, hidden_units]), scale=tf.ones([num_features, hidden_units]))
b_0 = Normal(loc=tf.zeros(hidden_units), scale=tf.ones(hidden_units))
W_1 = Normal(loc=tf.zeros([hidden_units, num_classes]), scale=tf.ones([hidden_units, num_classes]))
b_1 = Normal(loc=tf.zeros(num_classes), scale=tf.ones(num_classes))

X = tf.placeholder(tf.float32, [None, num_features])
y_ph_categorical = tf.placeholder(tf.int32, [None])
y_ph_multinomial = tf.placeholder(tf.float32, [None, num_classes])

logits = forward(X, W_0, b_0, W_1, b_1)
y = Multinomial(logits=logits, total_count=1.0)
y_cat = Categorical(logits=logits)

xs, ys = mnist.train.next_batch(5)
sess = ed.get_session()
# y.probs is basically 0 everywhere
print(sess.run([y.probs, y.log_prob(y_ph_multinomial)], {y_ph_multinomial: ys, X: xs}))
print(sess.run(y_cat.log_prob(y_ph_categorical), {y_ph_categorical: np.argmax(ys, 1), X: xs}))

# this works
x_data = np.zeros([2, 5], dtype=np.float32)
x_data[0, 0] = 1.0
x_data[1, 2] = 1.0
x = Multinomial(logits=tf.zeros([2, 5]), total_count=1.0)
sess = ed.get_session()
sess.run(x.log_prob(x_data))
```

**Comment by YiMX** [comment]
I created a PR to tensorflow https://github.com/tensorflow/tensorflow/pull/12485. log_prob uses a very old manner to calculate the minus cross entropy, nan comes from log(0). This can be fixed by adding a lower limit to probs like 1e-8 something or use tf.nn.softmax_cross_entropy_with_logits instead.

**Comment by YiMX** [comment]
This bug has been fixed.
